### ISSUE TITLE
Introducing namespace to function system

### ISSUE THREAD

**Issue by jihoonson** [hyunsik/tokamak#2]
Tokamak is designed to support a wide range of applications with pluggable function system. As the Tokamak's community grows, I believe that a lot of plugins will also be introduced to Tokamak, which can cause many duplication in function names among different plugins. For example, two kinds of ```+``` operators are possible, each for single decimal values and matrices.

To alleviate this problem, I think that using namespace is a good solution, and believe that adding namespace is much easier when there are only small number of plugins. 

Currently, I'm concerned with the representation of functions with its namespace. For example, the ```+``` function of the decimal arithmetic namespace can be represented ```decimal::+```. Do you have any good idea?

**Comment by hyunsik** [comment]
The idea sounds good to me.

**Comment by jihoonson** [comment]
Thanks @hyunsik. Honestly, I've already started coding.
If you don't mind, may I take this ticket?

**Comment by hyunsik** [comment]
Sure, I've assigned it to you. Since Tokamak is still very early stage, PluginManager and FuncRegistry will be changed a lot for a while. It is only my concern.

**Comment by jihoonson** [comment]
Thanks. I understand your concern. I'll submit my patch ASAP. 
Currently, PluginManager consists of four flat components, i.e., ```packages```, ```types```, ```functions```, and ```input sources```. IMO, this structure needs to be changed to have a hierarchical components as following:

* packages
 * types
 * functions
 * input sources

With this hierarchical structure, packages can be regarded as namespace for types, functions, and input sources. (Input sources also have their name, so it would be good if we support namespace.)

Do you have any comment on this way?

**Comment by hyunsik** [comment]
That looks good to me. I agree with the namespace concept.

In terms of the hierarchy, I think that the flat scheme is fine now because:
* Even if we support namespace, users rarely use namespaces in their queries only when they need to resolve name conflicts.
  * It implies that symbol resolution will be mostly done without namespace. So, the flat scheme is simple and good to keep one HashMap for them.
* Once loaded, plugins will not unloaded until instance shutdown. In other words, they are immutable.
  * So, we don't need to maintain plugins after we build the symbol tables at startup.

**Comment by jihoonson** [comment]
I agree, but have a question. What is your most concern on this? Maybe it is the name resolution speed, right?

**Comment by hyunsik** [comment]
Yes, one reason is speed. As you know, however, the speed will be improved if we keep multiple HashMaps for both name and namespace + name.

My reason is the complexity compared to benefit. Resolving a symbol without any namespace should look for all packages. Otherwise, we need to maintain two HashMaps like I mentioned above, and we need to look up two hash maps. Hence, I thought the hierarchy will increase the complexity of code and data, like Schema implementation in Tajo. In contrast, the benefit is questionable. If we need to manage function, types, sources along with each package, the hierarchy may be better. However, they will be just immutable maps.

**Comment by jihoonson** [comment]
Thanks @hyunsik!
I've submitted my patch at #6.
